dataflowAPI/src/liveness.C

```diff
@@ -526,43 +526,34 @@ ReadWriteInfo LivenessAnalyzer::calcRWSets(Instruction curInsn, Block *blk, Addr
       }
     }
   }
-  InsnCategory category = curInsn.getCategory();
-  switch(category)
-  {
-  case c_CallInsn:
+  if(curInsn.isCall()) {
       // Call instructions not at the end of a block are thunks, which are not ABI-compliant.
       // So make conservative assumptions about what they may read (ABI) but don't assume they write anything.
       ret.read |= (abi->getCallReadRegisters());
       if(blk->lastInsnAddr() == a)
       {
           ret.written |= (abi->getCallWrittenRegisters());
       }
-    break;
-  case c_ReturnInsn:
+  }
+  else if(curInsn.isReturn()) {
     ret.read |= (abi->getReturnReadRegisters());
     // Nothing written implicitly by a return
-    break;
-  case c_BranchInsn:
+  }
+  else if(curInsn.isBranch()) {
     if(!curInsn.allowsFallThrough() && isExitBlock(blk))
     {
       //Tail call, union of call and return
-      ret.read |= ((abi->getCallReadRegisters()) |
-		   (abi->getReturnReadRegisters()));
+      ret.read |= ((abi->getCallReadRegisters()) | (abi->getReturnReadRegisters()));
       ret.written |= (abi->getCallWrittenRegisters());
     }
-    break;
-  default:
-    {
-      const bool isInterrupt = Dyninst::InstructionAPI::isSoftwareInterrupt(curInsn);
-      const bool isSyscall = Dyninst::InstructionAPI::isSystemCall(curInsn);
-
-      if (isInterrupt || isSyscall) {
-	ret.read |= (abi->getSyscallReadRegisters());
-	ret.written |= (abi->getSyscallWrittenRegisters());
-      }
+  }
+  else {
+    auto is_interrupt = Dyninst::InstructionAPI::isSoftwareInterrupt(curInsn);
+    if (is_interrupt || curInsn.isSyscall()) {
```

Comment:
For uniformity, any reason not to add a `isSoftwareInterrup()` method to the instruction?

```suggestion
    if (curInsn. isSoftwareInterrupt() || curInsn.isSyscall()) {
```

Reply:
There isn't a category for interrupts. I plan to update that next.

Reply:
Sure, the `isSoftwareInterrupt` method can be added when you add it as a category instead of now.